### PR TITLE
Shell - using shell_fprintf from multiple contexts

### PR DESCRIPTION
**Pull request by nordic-krch**
This patch allows to use shell_fprintf from other contexts than command handler. However, it requires command to indicate that shell should keep command context in foreground (it is done by command calling `shell_command_enter()`). Shell on that error code will not print prompt and will not accept any input from the user until command is terminated with CTRL+C or `shell_command_exit()` call.

It can be used for `bt scan on` command to start scanning and get devices printed on shell until it is terminated. Note that scanning is still enabled only context is returned to shell, allowing user to type new commands (eg `scan off`).

I see following next steps:
- add macro for creating foreground command with termination handler called on termination sequence (e.g. `SHELL_FG_CMD(scan, ...., cmd_scan_on, cmd_scan_off) where scan_off function would be called on termination).
- further in the future provide console interface through shell that could be passed to the command. Command then could use input and output through shell while staying in foreground. That could be used for starting another shell-type application like python.

Additional commits:
- Added foreground example command to shell sample.
- removed checkpatch errors from shell.c

#11604 will use that for scanning so it needs to go first.

### CODE REVIEW

**Comment by nordic-krch** [comment]
hang on, will rebase.

**Comment by Vudentz** [comment]
I guess -EINPROGRESS is probably a better return to indicate the command is asynchronous and I suggest adding a function to flag the execution complete.

The ctrl+c is probably a great idea to interrupt it, but it may be useful to put the command in background perhaps by introducing ctrl+z bg and fg, obviously this will make the shell a lot more complex.

**Comment by nordic-krch** [comment]
@Vudentz 
>  -EINPROGRESS is probably a better

Good point.

> introducing ctrl+z

For sure we will get there 😄

Meanwhile, I've added `shell_command_exit` call that will be used in `bt init` in `bt_ready` callback.

**Review by jakub-uC — CHANGES_REQUESTED** [review]
Some minor drawbacks.

**Comment by nordic-krch** [comment]
@jarz-nordic i've applied your comments.

**Review by jakub-uC — APPROVED** [review]

**Comment by nordic-krch** [comment]
Hold it for now. I've noticed that keeping command in foreground context by returning -EINPROGRESS is not an option because context may get  switched before command completes. I will add explicit call `shell_command_enter`.

**Comment by nordic-krch** [comment]
Added explicit call `shell_command_enter` to be called to enter foreground. Note that shell will keep command in foreground only if command return 0.

File: doc/subsystems/shell/shell.rst [hunk 1]
```diff
@@ -235,13 +235,39 @@ Simple command handler implementation:
...
 	}
 
-.. warning::
-	Do not use function :cpp:func:`shell_fprintf` or shell print macros:
-	:c:macro:`shell_print`, :c:macro:`shell_info`, :c:macro:`shell_warn`,
-	:c:macro:`shell_error` outside of the command handler because this
-	might lead to incorrect text display on the screen. If any text should
-	be displayed outside of the command context, then use the
-	:ref:`logger` or `printk` function.
+:cpp:func:`shell_fprintf` or shell print macros: :c:macro:`shell_print`,
```
Comment (dbkinder) [inline doc/subsystems/shell/shell.rst]: PR #11604 is making the same changes to this shell.rst file?  (I've commented on that PR).

**Review by dbkinder — CHANGES_REQUESTED** [review]
Doc changes here are also in PR #11604 ?

File: doc/subsystems/shell/shell.rst (hunk 1 above)
Comment (nordic-krch) [reply doc/subsystems/shell/shell.rst]: applied you suggestions from #11604.

File: subsys/shell/shell.c [hunk 2]
```diff
@@ -1193,6 +1216,35 @@ static int shell_instance_init(const struct shell *shell, const void *p_config,
...
+	}
+}
+
+static void kill_handler(const struct shell *shell)
+{
+	(void)shell_instance_uninit(shell);
+	k_thread_abort(k_current_get());
+}
+
+static void command_exit_handler(const struct shell *shell)
```
Comment (jarz-nordic) [inline subsys/shell/shell.c]: this can be `inline` function.

File: subsys/shell/shell.c [hunk 3]
```diff
@@ -1193,6 +1216,35 @@ static int shell_instance_init(const struct shell *shell, const void *p_config,
...
+
+	k_poll_signal_check(signal, &set, &res);
+
+	if (set) {
+		k_poll_signal_reset(signal);
+		handler(shell);
+	}
+}
+
+static void kill_handler(const struct shell *shell)
```
Comment (jarz-nordic) [inline subsys/shell/shell.c]: this can be inline function.

**Review by dbkinder — APPROVED** [review]

**Comment by jarz-nordic** [comment]
@nashif or @carlescufi : May you please merge it? I need this to finish SMP example.

**Pull request merged**